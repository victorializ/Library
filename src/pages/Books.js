import React, { useState }  from 'react';

import { 
  BooksList, 
  BooksTopLevelMenu as TopLevelMenu, 
  LoadingComponent 
} from '../components';
import { book } from '../types';
import { compose, compare } from '../utils/utils';
import { getAllBooks } from '../services/http-client';
import { useRequest } from '../hooks';
import { formatData } from '../services/format-data/books'; 

function Books() {

  const [ input, setInput ] = useState('');
  const [ sort, setSort ] = useState('');
  const [ filter, setFilter ] = useState(book.Name);

  const response = useRequest(getAllBooks);
  
  const sortFunction = arr => 
    !sort ? arr : 
      arr.sort((a, b) => compare(a, b, sort, filter));
  
  const findFunction = arr => 
    !input || !filter ? arr :
      arr.filter(book => formatData(book)[filter].toString().toLowerCase().includes(input.toLowerCase()));

  const putInOrder = items => compose(findFunction, sortFunction)(items);

  return (
    <>
      <TopLevelMenu 
        sort={sort}
        input={input}
        filter={filter}
        onSortSet={sort => setSort(sort)} 
        onSortUnsert={() => setSort('')}
        onInput={input => setInput(input)}
        onFilterSet={filter => setFilter(filter)} 
        onFilterUnsert={() => setFilter('')}
      />
      <div className="books__list">
        <LoadingComponent {...response}>
          <BooksList data={putInOrder(response.data)}/>
        </LoadingComponent> 
      </div>
    </>
  );
}

export default Books;