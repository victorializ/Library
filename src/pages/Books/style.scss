@import '../../assets/variables.scss';

.books {
    &__label {
        height: 50px;
    }
    &__menu {
        margin: 0;
        height: 30vh;
        @include column-flex-container;
        background-image: $main-background;
    }
    &__search, &__filter {
        @include centered-horizontal-container;
        margin: $space-xs;
    }
}

.ui.search {
    margin: $space-sm;
}