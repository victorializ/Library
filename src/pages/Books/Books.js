import React,  { useState, useEffect }  from 'react';

import { Search, Button } from 'semantic-ui-react'

import { constants } from '../../i18n';
import { book, sorting } from '../../types';
import { colors } from '../../assets/semantic-colors';
import { formatData } from '../../services/format-data/books';
import { getAllBooks } from '../../services/http-client/books';

import './style.scss';
import { BooksList } from '../../components';

const MenuButton = ({name, selectedButton, clickHandler}) => {
  const getButtonColor = (buttonName, selected) =>  {
    return buttonName === selected ? colors.primary : colors.grey;
  }
  return (
    <Button 
      onClick={() => clickHandler(name)}
      color={getButtonColor(name, selectedButton)}>
        {constants[name]}
    </Button>
  )
}

const TopLevelMenu = ({filter, onFilterSet, input, onInput, sort, onSortSet}) =>
  <div className="books__filters">
    <div className="books__search">
      <Button.Group>
        <Button color={colors.secondary}>{constants.filter}</Button>
        <MenuButton name={book.name} selectedButton={filter} clickHandler={onFilterSet} />
        <Button.Or />
        <MenuButton name={book.bookYear} selectedButton={filter} clickHandler={onFilterSet} />
        <Button.Or />
        <MenuButton name={book.bookAuthors} selectedButton={filter} clickHandler={onFilterSet} />
        <Button.Or />
        <MenuButton name={book.bookGenres} selectedButton={filter} clickHandler={onFilterSet} />
      </Button.Group>
      <Search
        className="books__list"
        value = {input}
        onSearchChange ={(e, {value}) => onInput(value)}
      ></Search>
    </div>
    <Button.Group>
      <Button color={colors.secondary}>{constants.sort}</Button>
      <MenuButton name={sorting.asc} selectedButton={sort} clickHandler={onSortSet} />
      <Button.Or />
      <MenuButton name={sorting.desc} selectedButton={sort} clickHandler={onSortSet} />
    </Button.Group>
</div>


function Books() {
  const [ input, setInput ] = useState('');
  const [ sort, setSort ] = useState('');
  const [ filter, setFilter ] = useState(book.Name);
  const [ books, setBooks ] = useState([]);

  useEffect(() => {
    getAllBooks().then(res => setBooks(res.data.map(element => formatData(element))));
  }, []);


  const putInOrder = (items) => {
    const fileredItems = input === '' ? 
      items : items.filter(book => book[filter].toString().includes(input));
    const sortedItems = sort === '' ? 
    fileredItems : fileredItems.sort((a, b) => 
      sort === sorting.asc ?
        a[filter] < b[filter] ? -1 : 1 : 
        a[filter] > b[filter] ? -1 : 1);
    return sortedItems;
  }

  return (
    <div className="books__wrapper">
      <TopLevelMenu 
        sort={sort}
        input={input}
        filter={filter}
        onSortSet={sort => setSort(sort)} 
        onInput={input => setInput(input)}
        onFilterSet={filter => setFilter(filter)} 
      />
      <BooksList elements={putInOrder(books)}></BooksList>
    </div>
  );
}

export { Books };