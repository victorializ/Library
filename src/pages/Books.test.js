import React from 'react';
import { shallow } from 'enzyme';

import Books from './Books';

describe('Books page component', () => {
    it('renders correctly', () => {
        expect(shallow(<Books />)).toMatchSnapshot();
    }); 
});