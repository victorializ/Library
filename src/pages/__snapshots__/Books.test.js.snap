// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Books page component renders correctly 1`] = `
<Fragment>
  <TopLevelMenu
    input=""
    onFilterSet={[Function]}
    onFilterUnsert={[Function]}
    onInput={[Function]}
    onSortSet={[Function]}
    onSortUnsert={[Function]}
    sort=""
  />
  <div
    className="books__list"
  >
    <LoadingComponent
      data={null}
      error={
        Object {
          "message": null,
        }
      }
      loading={true}
    >
      <List
        data={null}
      />
    </LoadingComponent>
  </div>
</Fragment>
`;
